### ISSUE TITLE
Code improvements to xmakeglyphfontspecs()

### ISSUE THREAD

**Issue by veltza** [cog1to/st-ligatures#40]
I have a couple of suggestions for improvements to the xmakeglyphfontspecs function that you might be interested in.

Since version 0.9, the xmakeglyphfontspecs calls have been moved inside a loop in the xdrawline function. This ensures that all glyphs have the same attributes and colors when passed to the xmakeglyphfontspecs function:

```c
		if ((i > 0) && ATTRCMP(base, new)) {
			numspecs = xmakeglyphfontspecs(specs, &line[ox], x - ox, ox, y1);
			xdrawglyphfontspecs(specs, base, numspecs, ox, y1, x - ox);
			i = 0;
		}
```

So there is no need to check them again in the xmakeglyphfontspec function:

```c
		if (
			prevmode != mode
			|| ATTRCMP(glyphs[start], glyphs[i])
			|| selected(x + i, y) != selected(x + start, y)
			|| i == (len - 1)
		) {
```

So, for this reason, the above check and the entire outer loop can be removed from that function, which not only simplifies the implementation, but also fixes at least one edge case bug there (*).

But that's not all. Since all glyphs have the same attributes, it also simplifies the boxdraw variant, because you can take the box-drawing code out of the inner loop and handle box-drawing characters without feeding them to HarfBuzz. For example, this way:

```c
	if (glyphs[0].mode & ATTR_BOXDRAW) {
		xp = winx, yp = winy + font->ascent;
		for (numspecs = 0; numspecs < len; numspecs++, xp += runewidth) {
			/* minor shoehorning: boxdraw uses only this ushort */
			specs[numspecs].font = font->match;
			specs[numspecs].glyph = boxdrawindex(&glyphs[numspecs]);
			specs[numspecs].x = xp;
			specs[numspecs].y = yp;
		}
		return numspecs;
	}
```

I think these improvements would make this excellent patch even better.

Cheers

---

(*) Here's an example of the edge case bug:

1. Print the following text with `printf '\033[7mE\033[0mdge-case\n'`

![1](https://github.com/user-attachments/assets/463a835d-5b51-4318-8cc1-f25b128529b5)

2. Select the letter `E` with the mouse and you should see how the letters after the selection have moved to the left.

![2](https://github.com/user-attachments/assets/ed0a217c-2286-4855-9aab-96ce5e0cb1d1)

---

Edit: added an example of the edge case bug.

**Comment by cog1to** [comment]
I checked the `st` [code](http://git.suckless.org/st/file/x.c.html) and seems like `xmakeglyphfontspecs` call is still outside the main loop in `xdrawline`:

```
void
xdrawline(Line line, int x1, int y1, int x2)
{
  int i, x, ox, numspecs;
  Glyph base, new;
  XftGlyphFontSpec *specs = xw.specbuf;

  numspecs = xmakeglyphfontspecs(specs, &line[x1], x2 - x1, x1, y1);
  i = ox = 0;
  for (x = x1; x < x2 && i < numspecs; x++) {
    ...
```

are you looking at some kind of fork or a patch that makes that change? the change you're suggesting here makes sense, but it doesn't match with what I see in the official repo.

**Comment by cog1to** [comment]
oh wait, you were referring to my code 😅

**Comment by cog1to** [comment]
Ok, the first suggestion seems good.

I don't see much sense in the second one though. I get that we can avoid harfbuzz for boxdraw segments, but unless there's a serious performance concerns, I'd like to keep it inside the loop for consistency.

**Comment by veltza** [comment]
No, there are no serious performance concerns as there are usually very few box drawing characters on the screen. I just thought it would simplify the boxdraw variant because then the loop would stay the same in all variants. But it's your decision and I'm fine with it.

However, there are still a couple of things in your new code that you might want to clean up.


```c
	/* initial values. */
	mode = prevmode = glyphs[0].mode & ~attr_wrap;
	xresetfontsettings(mode, &font, &frcflags);
```

The prevmode variable is no longer needed. You could also pass the glyph mode directly to the function and get rid of the both mode variables.

```c
		/* Advance the drawing cursor if we've moved to a new cluster */
		if (code_idx > 0 && idx != shaped.glyphs[code_idx - 1].cluster) {
			xp += runewidth;
			cluster_xp = xp;
			cluster_yp = yp;
			runewidth = win.cw * ((glyphs[idx].mode & ATTR_WIDE) ? 2.0f : 1.0f);
		}
```

The runewidth does not change in the loop because all glyphs are either wide glyphs or not. So you don't need to recalculate it.

```c
			/* Nothing was found. Use fontconfig to find matching font. */
			if (f >= frclen) {
				if (!font->set)
					font->set = FcFontSort(0, font->pattern,
																 1, 0, &fcres);
				fcsets[0] = font->set;
```

There are too many tabs in the FcFontSort function call, so the code may go off the screen.

**Comment by cog1to** [comment]
Thanks for the review, I fixed the issues.

Regarding passing `mode` directly to the function, I don't think it's a very good idea for compatibility reasons. This is a patch after all, and changing function signatures likely will cause other patches to break. I try to keep such changes to a minimum.
